File: sol/368/368.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        n = len(nums)
        nums.sort()
        chainLen = [1] * n
        prevIdx = [-1] * n
        maxLen = 0
        maxIdx = 0
        for i in range(n):
            for j in range(i + 1, n):
                if nums[j] % nums[i] != 0:
                    continue
                if chainLen[j] < chainLen[i] + 1:
                    chainLen[j] = chainLen[i] + 1
                    prevIdx[j] = i
                    if chainLen[j] > maxLen:
                        maxLen = chainLen[j]
                        maxIdx = j
        res = []
        while maxIdx != -1:
            res.append(nums[maxIdx])
            maxIdx = prevIdx[maxIdx]
        return res

File: sol/368/368.cpp
struct Solution {
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int n = nums.size();
        ranges::sort(nums);
        vector<int> chainLen(n, 1), prevIdx(n, -1);
        int maxLen = 0, maxIdx = 0;
        for (int i = 0; i < n; ++i){
            for (int j = i + 1; j < n; ++j){
                if (nums[j] % nums[i] != 0){
                    continue;
                }
                if (chainLen[j] < chainLen[i] + 1){
                    chainLen[j] = chainLen[i] + 1;
                    prevIdx[j] = i;
                    if (chainLen[j] > maxLen){
                        maxLen = chainLen[j];
                        maxIdx = j;
                    }
                }
            }
        }
        vector<int> res;
        while (maxIdx != -1){
            res.push_back(nums[maxIdx]);
            maxIdx = prevIdx[maxIdx];
        }
        return res;
    }
};
